src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -684,6 +688,22 @@ std::string ClusterTaskManager::DebugStr() const {
   return buffer.str();
 }
 
+void ClusterTaskManager::RecordMetrics() {
+  stats::NumQueuedTasks.Record(metric_tasks_queued_);
+  stats::NumDispatchedTasks.Record(metric_tasks_dispatched_);
+  stats::NumSpilledTasks.Record(metric_tasks_spilled_);
+
+  metric_tasks_queued_ = 0;
+  metric_tasks_dispatched_ = 0;
+  metric_tasks_spilled_ = 0;
```

Comment:
Prometheus counts are supposed to be monotonically nondecreasing, but these counts will oscillate between recordings given that you’re resetting the counts to 0 after each report. Unless I’m missing some interaction between OpenCensus measure views and the Prometheus exporter, those counts should instead grow unbounded, and will be wrapped to 0 upon overflow, which Prometheus will see as a restart of the count and handle as expected.

Reply:
Interesting, I based my approach of of the existing code like this: https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_server/gcs_worker_manager.cc#L57

which looks like it's only trying to record the delta. Are you saying that's buggy?

Reply:
It was my understanding that `.Record()` was for recording the actual value of the metric, which is useful for e.g. counts that are very frequently updated but we only wish to actually record periodically, instead of the deltas, which are supposed to be recorded with dedicated delta functions like `.Inc()`. But this is off a quick skim of the OpenCensus layer a while ago, I'll double-check that.

Reply:
Oh boy, it actually appears to be neither. Count metrics use the `Count` aggregation, which counts the number of _records_, ignoring the individual values of each recording. If we want to record deltas, we should be using the `Sum` aggregation, and if we want to use cumulative count recordings, we should be using the `LastValue` aggregation.

https://github.com/census-instrumentation/opencensus-cpp/blob/master/opencensus/stats/aggregation.h